compiler/plc_xml/src/model/fbd.rs

```diff
@@ -105,14 +122,239 @@ impl Parseable for FunctionBlockDiagram {
+                if let Some(ref_id) = conn.ref_local_id {
+                    resolve_sink_connection_if_needed(*id, ref_id, None)
+                }
+            }
+        });
+
+        (resolved_connections, diagnostics)
+    }
+
+    /// Attempts to resolve the LValue or RValue of a sink connection-point by checking the associated node referenced in the
+    /// provided `source_connection` HashMap. Should the associated source point to another sink (i.e. not an assignable value),
+    /// this function will be called recursively until a valid value is found.
+    /// Additionally, cyclic connections are detected and reported.
+    fn find_assignable_sink_value(
+        &self,
+        source_connections: &HashMap<&str, SourceReference>,
+        node_id: NodeId,
+        connector_path: &mut IndexSet<String>,
+        source_location_factory: &SourceLocationFactory,
+    ) -> Result<NodeId, Diagnostic> {
+        self.get(&node_id)
+            .map(|node| {
+                let connector_name = node.get_name();
+                let Some(source) = source_connections.get(connector_name.as_str()) else {
+                    return Err(Diagnostic::sink_without_associated_source(
+                        &connector_name,
+                        source_location_factory.create_block_location(node.get_id(), node.get_exec_id()))
+                    )
+                };
+
+                match source {
+                    SourceReference::Value(id) => Ok(*id),
+                    // for direct sink-to-source connections, we need to recurse to find the actual value
+                    SourceReference::Connector(id) => {
+                        if connector_path.insert(connector_name) {
+                            self.find_assignable_sink_value(
+                                source_connections,
+                                *id,
+                                connector_path,
+                                source_location_factory,
+                            )
+                        } else {
+                            // data-recursion detected
+                            let mut message = String::new();
+                            for name in connector_path.iter() {
+                                message = format!("{message}{name} -> ")
+                            }
+
+                            message = format!("{}{}", message, node.get_name());
+
+                            Err(Diagnostic::cyclic_connection(
+                                message,
+                                source_location_factory.create_block_location(*id, None),
+                            ))
+                        }
+                    }
+                    SourceReference::Unconnected(id) => Err(Diagnostic::unconnected_source(
+                        &connector_name,
+                        source_location_factory.create_block_location(*id, None),
+                    )),
+                }
+            })
+            .unwrap()
```

Comment:
We could probably avoid this as well

Reply:
Similar thought to above - the id passed into this function is taken from an existing node, so the `get` method will never return none. I agree that this is not really made clear at all, maybe a let-else would be better here.